_posts/2022-08-04-calldata-tuple-reencoding-head-overflow-bug.md

```diff
@@ -0,0 +1,274 @@
+A similar situation happens inside `T` - when `y` is being written the data belonging to the tail of `x` is already in place.
+
+## Effects of the Bug
+
+The aggressive cleanup combined with the non-linear encoding order of dynamic types resulted in a bug where
+the initial 32 bytes of the tail could be overwritten during the cleanup of the last head component.
+
+The dynamic tuple example above meets all the criteria for the bug so introducing ABI reencoding would trigger it:
+
+```solidity
+struct T {
+    bytes x;
+    uint[3] y;
+}
+
+contract E {
+    function f(bool a, T calldata b, bytes32[2] calldata c)
+        public
+        returns (bool, T calldata, bytes32[2] calldata)
+    {
+        return (a, b, c);
+    }
+}
+```
+
+Even though `E.f()` is meant to simply return the same input it receives, due to the bug it would produce faulty encoding:
+
+```
+0x0000000000000000000000000000000000000000000000000000000000000001 a
+0x0000000000000000000000000000000000000000000000000000000000000080 offset of b
+0x6100000000000000000000000000000000000000000000000000000000000000 c[0]
+0x6200000000000000000000000000000000000000000000000000000000000000 c[1]
+0x0000000000000000000000000000000000000000000000000000000000000000 offset of b.x       // Wrong value
+0x000000000000000000000000000000000000000000000000000000000000000b b.y[0]
+0x000000000000000000000000000000000000000000000000000000000000000c b.y[1]
+0x000000000000000000000000000000000000000000000000000000000000000d b.y[2]
+0x0000000000000000000000000000000000000000000000000000000000000000 b.x (length field)  // Wrong value
+0x6162636400000000000000000000000000000000000000000000000000000000 b.x (data)
+```
+
+Here, the cleanup performed for `b.y` would overwrite the length field of `b.x`.
+Similarly, the cleanup of `c` would overwrite the offset of `b.x`.
+
+The case with a static tuple was not affected because, in absence of the tail, the encoding order is linear.
+The encoder could still write past the end of the whole area reserved for the encoding but the compiler would be able
+to ensure that the memory past it had not been allocated for another purpose, making the cleanup perfectly safe.
+
+The bug only affected the encoding of tuples in the presence of a calldata array because that was the only situation where
+the routine that performs the aggressive cleanup was used.
+The routine is a part of the IR-based generator and runs regardless of the use of the optimizer.
+Since the legacy code generation pipeline does not have its own ABI coder v2 implementation and uses
+the IR-based generator for ABI encoding, the bug can be triggered when using either of them.
+
+## Impact
+
+The bug is very easy to trigger in contrived situations but requires a combination of multiple factors to manifest.
+This seems to lower the likelihood of it being triggered, given that it remained undetected through several major releases.
+It may however have gone undetected in the encoded data of events or errors.
```

Comment:
I mean it's common to emit events with data you just might need even if you don't really use it. And with errors, if decoding a revert causes another revert that also might go unnoticed.

But yeah, it's a bit of a stretch, I just thought it might be worth highlighting that this may occur in events and errors too and it's harder to notice there.

Reply:
Yeah, fair enough.